File: laba3/cod.java
1. Бинарная куча
java
public class BinaryHeap {
    private int[] heap;
    private int size;
    private int capacity;
    
    public BinaryHeap(int capacity) {
        this.capacity = capacity;
        this.heap = new int[capacity];
        this.size = 0;
    }
    
    private int parent(int i) { return (i - 1) / 2; }
    private int left(int i) { return 2 * i + 1; }
    private int right(int i) { return 2 * i + 2; }
    
    private void swap(int i, int j) {
        int temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }
    
    private void heapifyUp(int i) {
        while (i > 0 && heap[i] < heap[parent(i)]) {
            swap(i, parent(i));
            i = parent(i);
        }
    }
    
    private void heapifyDown(int i) {
        int smallest = i;
        int left = left(i);
        int right = right(i);
        
        if (left < size && heap[left] < heap[smallest])
            smallest = left;
        if (right < size && heap[right] < heap[smallest])
            smallest = right;
        
        if (smallest != i) {
            swap(i, smallest);
            heapifyDown(smallest);
        }
    }
    
    public void insert(int key) {
        if (size == capacity) {
            System.out.println("Heap is full");
            return;
        }
        
        heap[size] = key;
        heapifyUp(size);
        size++;
    }
    
    public int extractMin() {
        if (size == 0) throw new IllegalStateException("Heap is empty");
        
        int min = heap[0];
        heap[0] = heap[size - 1];
        size--;
        heapifyDown(0);
        return min;
    }
    
    public int getMin() {
        if (size == 0) throw new IllegalStateException("Heap is empty");
        return heap[0];
    }
    
    public boolean isEmpty() { return size == 0; }
    public int size() { return size; }

    // Тестирование
    public static void main(String[] args) {
        BinaryHeap heap = new BinaryHeap(10);
        heap.insert(5);
        heap.insert(3);
        heap.insert(8);
        heap.insert(1);
        
        System.out.println("Min: " + heap.getMin());  // 1
        System.out.println("Extract min: " + heap.extractMin());  // 1
        System.out.println("New min: " + heap.getMin());  // 3
    }
}
2.куча фибоначи
  public class FibonacciHeap<T extends Comparable<T>> {
    private static class Node<T> {
        T key;
        int degree = 0;
        Node<T> parent;
        Node<T> child;
        Node<T> left = this;
        Node<T> right = this;
        boolean marked = false;
        
        Node(T key) {
            this.key = key;
        }
    }
    
    private Node<T> minNode;
    private int size;
    
    public boolean isEmpty() {
        return minNode == null;
    }
    
    public Node<T> insert(T key) {
        Node<T> newNode = new Node<>(key);
        
        if (minNode == null) {
            minNode = newNode;
        } else {
            // Add to root list
            newNode.left = minNode.left;
            newNode.right = minNode;
            minNode.left.right = newNode;
            minNode.left = newNode;
            
            if (key.compareTo(minNode.key) < 0) {
                minNode = newNode;
            }
        }
        
        size++;
        return newNode;
    }
    
    public T getMin() {
        if (minNode == null) throw new IllegalStateException("Heap is empty");
        return minNode.key;
    }
    
    public T extractMin() {
        if (minNode == null) throw new IllegalStateException("Heap is empty");
        
        Node<T> min = minNode;
        T minKey = min.key;
        
        // Add children to root list
        if (min.child != null) {
            Node<T> child = min.child;
            Node<T> firstChild = child;
            do {
                Node<T> nextChild = child.right;
                // Remove from child list
                child.left.right = child.right;
                child.right.left = child.left;
                // Add to root list
                child.left = minNode.left;
                child.right = minNode;
                minNode.left.right = child;
                minNode.left = child;
                child.parent = null;
                child = nextChild;
            } while (child != firstChild);
        }
        
        // Remove min from root list
        min.left.right = min.right;
        min.right.left = min.left;
        
        if (min == min.right) {
            minNode = null;
        } else {
            minNode = min.right;
            consolidate();
        }
        
        size--;
        return minKey;
    }
    
    private void consolidate() {
        if (minNode == null) return;
        
        int maxDegree = (int) (Math.log(size) / Math.log(2)) + 1;
        @SuppressWarnings("unchecked")
        Node<T>[] degreeTable = new Node[maxDegree + 1];
        
        java.util.ArrayList<Node<T>> roots = new java.util.ArrayList<>();
        Node<T> current = minNode;
        Node<T> first = current;
        
        do {
            roots.add(current);
            current = current.right;
        } while (current != first);
        
        for (Node<T> node : roots) {
            int degree = node.degree;
            while (degreeTable[degree] != null) {
                Node<T> other = degreeTable[degree];
                if (node.key.compareTo(other.key) > 0) {
                    Node<T> temp = node;
                    node = other;
                    other = temp;
                }
                
                link(other, node);
                degreeTable[degree] = null;
                degree++;
            }
            degreeTable[degree] = node;
        }
        
        minNode = null;
        for (Node<T> node : degreeTable) {
            if (node != null) {
                if (minNode == null) {
                    minNode = node;
                    node.left = node;
                    node.right = node;
                } else {
                    node.left = minNode.left;
                    node.right = minNode;
                    minNode.left.right = node;
                    minNode.left = node;
                    
                    if (node.key.compareTo(minNode.key) < 0) {
                        minNode = node;
                    }
                }
            }
        }
    }
    
    private void link(Node<T> child, Node<T> parent) {
        // Remove child from root list
        child.left.right = child.right;
        child.right.left = child.left;
        
        // Make child a child of parent
        child.parent = parent;
        if (parent.child == null) {
            parent.child = child;
            child.right = child;
            child.left = child;
        } else {
            child.left = parent.child.left;
            child.right = parent.child;
            parent.child.left.right = child;
            parent.child.left = child;
        }
        
        parent.degree++;
        child.marked = false;
    }
    
    public void decreaseKey(Node<T> node, T newKey) {
        if (newKey.compareTo(node.key) > 0) {
            throw new IllegalArgumentException("New key must be smaller");
        }
        
        node.key = newKey;
        Node<T> parent = node.parent;
        
        if (parent != null && node.key.compareTo(parent.key) < 0) {
            cut(node, parent);
            cascadingCut(parent);
        }
        
        if (node.key.compareTo(minNode.key) < 0) {
            minNode = node;
        }
    }
    
    private void cut(Node<T> node, Node<T> parent) {
        // Remove node from parent's child list
        if (node.right == node) {
            parent.child = null;
        } else {
            node.left.right = node.right;
            node.right.left = node.left;
            if (parent.child == node) {
                parent.child = node.right;
            }
        }
        
        parent.degree--;
        
        // Add node to root list
        node.left = minNode.left;
        node.right = minNode;
        minNode.left.right = node;
        minNode.left = node;
        
        node.parent = null;
        node.marked = false;
    }
    
    private void cascadingCut(Node<T> node) {
        Node<T> parent = node.parent;
        if (parent != null) {
            if (!node.marked) {
                node.marked = true;
            } else {
                cut(node, parent);
                cascadingCut(parent);
            }
        }
    }
    
    public void delete(Node<T> node) {
        decreaseKey(node, getMinValue());
        extractMin();
    }
    
    @SuppressWarnings("unchecked")
    private T getMinValue() {
        // Return minimal possible value for type T
        if (minNode.key instanceof Integer) {
            return (T) Integer.valueOf(Integer.MIN_VALUE);
        } else if (minNode.key instanceof Double) {
            return (T) Double.valueOf(Double.MIN_VALUE);
        } else {
            throw new UnsupportedOperationException("Unsupported type");
        }
    }
    
    public int size() { return size; }

    // Тестирование
    public static void main(String[] args) {
        FibonacciHeap<Integer> fibHeap = new FibonacciHeap<>();
        FibonacciHeap<Integer>.Node<Integer> node1 = fibHeap.insert(10);
        FibonacciHeap<Integer>.Node<Integer> node2 = fibHeap.insert(5);
        FibonacciHeap<Integer>.Node<Integer> node3 = fibHeap.insert(20);
        FibonacciHeap<Integer>.Node<Integer> node4 = fibHeap.insert(3);
        FibonacciHeap<Integer>.Node<Integer> node5 = fibHeap.insert(8);
        
        System.out.println("Min: " + fibHeap.getMin());  // 3
        System.out.println("Extract min: " + fibHeap.extractMin());  // 3
        System.out.println("New min: " + fibHeap.getMin());  // 5
        
        fibHeap.decreaseKey(node2, 2);
        System.out.println("After decrease key, min: " + fibHeap.getMin());  // 2
    }
}
3. Хеш-таблица
java
public class HashTable<K, V> {
    private static class Entry<K, V> {
        K key;
        V value;
        Entry<K, V> next;
        
        Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }
    
    private Entry<K, V>[] table;
    private int capacity;
    private int size;
    
    @SuppressWarnings("unchecked")
    public HashTable(int capacity) {
        this.capacity = capacity;
        this.table = new Entry[capacity];
        this.size = 0;
    }
    
    private int hash(K key) {
        return Math.abs(key.hashCode()) % capacity;
    }
    
    public void put(K key, V value) {
        if (size >= capacity * 0.7) resize();
        
        int index = hash(key);
        Entry<K, V> newEntry = new Entry<>(key, value);
        
        if (table[index] == null) {
            table[index] = newEntry;
        } else {
            Entry<K, V> current = table[index];
            while (current.next != null) {
                if (current.key.equals(key)) {
                    current.value = value;
                    return;
                }
                current = current.next;
            }
            if (current.key.equals(key)) {
                current.value = value;
            } else {
                current.next = newEntry;
            }
        }
        size++;
    }
    
    public V get(K key) {
        int index = hash(key);
        Entry<K, V> current = table[index];
        
        while (current != null) {
            if (current.key.equals(key)) {
                return current.value;
            }
            current = current.next;
        }
        throw new RuntimeException("Key not found: " + key);
    }
    
    public void remove(K key) {
        int index = hash(key);
        Entry<K, V> current = table[index];
        Entry<K, V> prev = null;
        
        while (current != null) {
            if (current.key.equals(key)) {
                if (prev == null) {
                    table[index] = current.next;
                } else {
                    prev.next = current.next;
                }
                size--;
                return;
            }
            prev = current;
            current = current.next;
        }
        throw new RuntimeException("Key not found: " + key);
    }
    
    @SuppressWarnings("unchecked")
    private void resize() {
        capacity *= 2;
        Entry<K, V>[] oldTable = table;
        table = new Entry[capacity];
        size = 0;
        
        for (Entry<K, V> entry : oldTable) {
            Entry<K, V> current = entry;
            while (current != null) {
                put(current.key, current.value);
                current = current.next;
            }
        }
    }
    
    public boolean contains(K key) {
        try {
            get(key);
            return true;
        } catch (RuntimeException e) {
            return false;
        }
    }
    
    public int size() { return size; }

    // Тестирование
    public static void main(String[] args) {
        HashTable<String, Integer> table = new HashTable<>(10);
        table.put("apple", 1);
        table.put("banana", 2);
        table.put("cherry", 3);
        
        System.out.println("apple: " + table.get("apple"));  // 1
        table.remove("banana");
        System.out.println("Contains banana: " + table.contains("banana"));  // false
    }
}

File: laba3/cod.py
Бинарная куча
class BinaryHeap:
    def __init__(self):
        self.heap = []
    
    def parent(self, i):
        return (i - 1) // 2
    
    def left_child(self, i):
        return 2 * i + 1
    
    def right_child(self, i):
        return 2 * i + 2
    
    def insert(self, key):
        self.heap.append(key)
        self._heapify_up(len(self.heap) - 1)
    
    def _heapify_up(self, i):
        while i > 0 and self.heap[i] < self.heap[self.parent(i)]:
            self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
            i = self.parent(i)
    
    def extract_min(self):
        if not self.heap:
            return None
        
        min_val = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        self._heapify_down(0)
        return min_val
    
    def _heapify_down(self, i):
        smallest = i
        left = self.left_child(i)
        right = self.right_child(i)
        
        if left < len(self.heap) and self.heap[left] < self.heap[smallest]:
            smallest = left
        
        if right < len(self.heap) and self.heap[right] < self.heap[smallest]:
            smallest = right
        
        if smallest != i:
            self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
            self._heapify_down(smallest)
    
    def get_min(self):
        return self.heap[0] if self.heap else None
    
    def is_empty(self):
        return len(self.heap) == 0

# Тестирование
heap = BinaryHeap()
heap.insert(5)
heap.insert(3)
heap.insert(8)
heap.insert(1)
print("Min:", heap.get_min())  # 1
print("Extract min:", heap.extract_min())  # 1
print("New min:", heap.get_min())  # 3
...................................................
куча фибоначи

class FibonacciNode:
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.parent = None
        self.child = None
        self.left = self
        self.right = self
        self.marked = False

class FibonacciHeap:
    def __init__(self):
        self.min_node = None
        self.node_count = 0
    
    def is_empty(self):
        return self.min_node is None
    
    def insert(self, key):
        new_node = FibonacciNode(key)
        
        if self.min_node is None:
            self.min_node = new_node
        else:
            # Добавляем новый узел в корневой список слева от min_node
            new_node.left = self.min_node.left
            new_node.right = self.min_node
            self.min_node.left.right = new_node
            self.min_node.left = new_node
            
            if key < self.min_node.key:
                self.min_node = new_node
        
        self.node_count += 1
        return new_node
    
    def get_min(self):
        return self.min_node.key if self.min_node else None
    
    def extract_min(self):
        if self.min_node is None:
            return None
        
        min_node = self.min_node
        min_key = min_node.key
        
        # Добавляем детей min_node в корневой список
        if min_node.child:
            child = min_node.child
            first_child = child
            while True:
                next_child = child.right
                # Добавляем ребенка в корневой список
                child.left.right = child.right
                child.right.left = child.left
                child.left = self.min_node.left
                child.right = self.min_node
                self.min_node.left.right = child
                self.min_node.left = child
                child.parent = None
                
                if next_child == first_child:
                    break
                child = next_child
        
        # Удаляем min_node из корневого списка
        min_node.left.right = min_node.right
        min_node.right.left = min_node.left
        
        if min_node == min_node.right:
            self.min_node = None
        else:
            self.min_node = min_node.right
            self._consolidate()
        
        self.node_count -= 1
        return min_key
    
    def _consolidate(self):
        if self.min_node is None:
            return
        
        # Создаем массив для деревьев разных степеней
        degree_table = [None] * (self.node_count.bit_length() + 1)
        
        nodes = []
        current = self.min_node
        first = current
        
        # Собираем все корневые узлы
        while True:
            nodes.append(current)
            current = current.right
            if current == first:
                break
        
        for node in nodes:
            degree = node.degree
            while degree_table[degree] is not None:
                other = degree_table[degree]
                if node.key > other.key:
                    node, other = other, node
                
                self._link(other, node)
                degree_table[degree] = None
                degree += 1
            
            degree_table[degree] = node
        
        # Восстанавливаем min_node
        self.min_node = None
        for node in degree_table:
            if node is not None:
                if self.min_node is None:
                    self.min_node = node
                    node.left = node
                    node.right = node
                else:
                    # Добавляем node в корневой список
                    node.left = self.min_node.left
                    node.right = self.min_node
                    self.min_node.left.right = node
                    self.min_node.left = node
                    
                    if node.key < self.min_node.key:
                        self.min_node = node
    
    def _link(self, child, parent):
        # Удаляем child из корневого списка
        child.left.right = child.right
        child.right.left = child.left
        
        # Делаем child дочерним для parent
        child.parent = parent
        if parent.child is None:
            parent.child = child
            child.left = child
            child.right = child
        else:
            child.left = parent.child.left
            child.right = parent.child
            parent.child.left.right = child
            parent.child.left = child
        
        parent.degree += 1
        child.marked = False
    
    def decrease_key(self, node, new_key):
        if new_key > node.key:
            raise ValueError("New key must be smaller than current key")
        
        node.key = new_key
        parent = node.parent
        
        if parent is not None and node.key < parent.key:
            self._cut(node, parent)
            self._cascading_cut(parent)
        
        if node.key < self.min_node.key:
            self.min_node = node
    
    def _cut(self, node, parent):
        # Удаляем node из списка детей parent
        if node.right == node:
            parent.child = None
        else:
            node.left.right = node.right
            node.right.left = node.left
            if parent.child == node:
                parent.child = node.right
        
        parent.degree -= 1
        
        # Добавляем node в корневой список
        node.left = self.min_node.left
        node.right = self.min_node
        self.min_node.left.right = node
        self.min_node.left = node
        
        node.parent = None
        node.marked = False
    
    def _cascading_cut(self, node):
        parent = node.parent
        if parent is not None:
            if not node.marked:
                node.marked = True
            else:
                self._cut(node, parent)
                self._cascading_cut(parent)
    
    def delete(self, node):
        self.decrease_key(node, float('-inf'))
        self.extract_min()

# Тестирование

fib_heap = FibonacciHeap()
nodes = []
nodes.append(fib_heap.insert(10))
nodes.append(fib_heap.insert(5))
nodes.append(fib_heap.insert(20))
nodes.append(fib_heap.insert(3))
nodes.append(fib_heap.insert(8))

print("Min:", fib_heap.get_min())  # 3
print("Extract min:", fib_heap.extract_min())  # 3
print("New min:", fib_heap.get_min())  # 5

fib_heap.decrease_key(nodes[1], 2)  # Уменьшаем ключ с 5 до 2
print("After decrease key, min:", fib_heap.get_min())  # 2

.............................
Хеш-таблица
class HashNode:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None

class HashTable:
    def __init__(self, size=10):
        self.size = size
        self.table = [None] * size
        self.count = 0
    
    def _hash(self, key):
        return hash(key) % self.size
    
    def put(self, key, value):
        if self.count / self.size > 0.7:
            self._resize()
        
        index = self._hash(key)
        new_node = HashNode(key, value)
        
        if self.table[index] is None:
            self.table[index] = new_node
        else:
            current = self.table[index]
            while current.next:
                if current.key == key:
                    current.value = value
                    return
                current = current.next
            if current.key == key:
                current.value = value
            else:
                current.next = new_node
        self.count += 1
    
    def get(self, key):
        index = self._hash(key)
        current = self.table[index]
        
        while current:
            if current.key == key:
                return current.value
            current = current.next
        raise KeyError(f"Key {key} not found")
    
    def remove(self, key):
        index = self._hash(key)
        current = self.table[index]
        prev = None
        
        while current:
            if current.key == key:
                if prev:
                    prev.next = current.next
                else:
                    self.table[index] = current.next
                self.count -= 1
                return
            prev = current
            current = current.next
        raise KeyError(f"Key {key} not found")
    
    def _resize(self):
        old_table = self.table
        self.size *= 2
        self.table = [None] * self.size
        self.count = 0
        
        for node in old_table:
            current = node
            while current:
                self.put(current.key, current.value)
                current = current.next

# Тестирование
ht = HashTable()
ht.put("apple", 1)
ht.put("banana", 2)
ht.put("cherry", 3)
print("Get apple:", ht.get("apple"))  # 1
ht.remove("banana")

